cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -125,7 +125,11 @@ func getInitConfigurationFromCluster(kubeconfigDir string, client clientset.Inte
 		// However, if newControlPlane == true, initcfg.NodeRegistration is not used at all and it's overwritten later on.
 		// Thus it's necessary to supply some default value, that will avoid the call to DetectCRISocket() and as
 		// initcfg.NodeRegistration is discarded, setting whatever value here is harmless.
-		initcfg.NodeRegistration.CRISocket = constants.DefaultDockerCRISocket
+		if criSocket != "" {
+			initcfg.NodeRegistration.CRISocket = criSocket
+		} else {
+			initcfg.NodeRegistration.CRISocket = constants.DefaultDockerCRISocket
+		}
```

Comment:
I think we can omit the else case here because the initconfiguration defaulting will already srt it to docker on empty.

Reply:
Emm... But I think we'd better keep `else` here because of the above comment.
Anyway, we should set `initcfg.NodeRegistration.CRISocket` as some value here to avoid running `DetectCRISocket()` before it should be run. But for `kubeadm join` case, we should set it to the correct value to avoid executing `docker info`.